Uninitialized variables

laBobberto opened the pull request:
Fixed: Procedure initialization to prevent undefined behavior.

Added initialization of variables with default values ​​when they are declared. This ensures predictable behavior and eliminates related bugs.

laBobberto committed:
fix: Initialize can_use_more to prevent compiler warning

The 'can_use_more' variable is now initialized with a default value at declaration. This resolves the "variable may be uninitialized" compiler warning that occurred when the initial 'length' was zero.

laBobberto committed:
fix: Initialize long_value to prevent potential errors

The 'long_value' variable was declared at the start of a loop without a default value. This created a risk of being used with a garbage value, triggering compiler warnings.

laBobberto committed:
fix: Initialize value variable to improve code robustness

This prevents potential use of an uninitialized variable, silences static analysis warnings, and makes the code safer and more predictable, even if the problematic code path was not reachable under the current logic.

laBobberto committed:
fix: Initialize 'length' variable in blr_print_dtype

The 'length' variable was declared without a default value in a function with complex control flow (two separate switch statements).

laBobberto committed:
fix: Fixed warning about uninitialized variable n

Initialized variable 'n' in blr_print_verb to 0 to avoid undefined behavior and remove compiler warning.

laBobberto committed:
fix: Initialize savNumber variable in BlockNode::execute

The 'savNumber' variable was not initialized when declared. This resulted in a compiler warning and risk of using a garbage value in 'case Request::req_unwind' if the transaction was a system one (TRA_system).

Added default initialization ({}) to ensure predictable behavior and fix a bug.

laBobberto committed:
fix: Initialize parentStream in EraseNode::pass1Erase

The 'parentStream' variable was not initialized when declared. This could lead to a random value being used on the second and subsequent loop iterations in the pass1Erase function.

Added default initialization ({}) to ensure safe behavior.

laBobberto committed:
fix: Initialize returnsPos in ExecBlockNode::genBlr

The 'returnsPos' variable was being set in two separate but logically mutually exclusive 'if' blocks. This was causing a false compiler warning about the possible use of an uninitialized variable.

Added initialization of 'returnsPos = 0' on declaration to eliminate the warning.

laBobberto committed:
fix: Initialize stream variables in ModifyNode::pass1Modify

The 'parentStream' and 'parentNewStream' variables were not initialized when declared. This caused the compiler to warn about possible use of their values before assignment on the second and subsequent iterations of the loop.

Added default initialization ({}) to both variables to ensure safe behavior.

laBobberto committed:
fix: Initialize parentStream variable in StoreNode::pass1Store

The 'parentStream' variable was not initialized when declared. This was causing a compiler warning.

Added default initialization ({}) to the variable to ensure safe behavior.

laBobberto committed:
fix: Initialize 'value' pointer in validateExpressions

The 'value' pointer was not initialized, which created a risk of dereferencing it with a random value in one of the branches of the conditional logic. This resulted in a compiler warning and potential undefined behavior.

Added initialization of the pointer to 'nullptr' to ensure a safe state.

laBobberto committed:
fix: Fix using uninitialized variable in stringToKey

In UnicodeUtil::Utf16Collation::stringToKey function, 'lastCharKeyLen' variable was left uninitialized if 'srcLenLong' was zero. This resulted in reading random value from memory and undefined behavior.

Added zero initialization for 'prefixLen' and 'lastCharKeyLen' when they are declared to fix the bug and eliminate compiler warning.

laBobberto committed:
fix: Initialize merge_pool pointer in Sort::sort method

The 'merge_pool' pointer was initialized only under the condition 'count > 1'. Although its reading was protected by the same condition, the static analyzer issued a warning about the possible use of an uninitialized variable.

Added initialization of the pointer to the value 'nullptr' when declaring.

laBobberto committed:
fix: Initialize 'value' variable in CVT_get_double

In CVT_get_double, the 'value' variable was left uninitialized if the 'default' block was executed in the switch statement. This caused the compiler to warn about the possible use of the variable before it was initialized.

Added 'value = 0.0' initialization on declaration to ensure safe behavior in all code paths.

laBobberto committed:
fix: Initialize ret_code variable in main loop

The 'ret_code' variable was not initialized when declared in the do-while loop.

Added initialization 'ret_code = 0' to make the code more robust.

laBobberto committed:
fix: Initialize minlen variable in EXE_assignment

The 'minlen' variable was left uninitialized if the data type in 'to_desc' was not handled by any of the 'case' in the switch block. This resulted in reading a random value and undefined behavior.

Added initialization 'minlen = 0' on declaration to fix the bug.

laBobberto committed:
fix: Initialize select_node pointer in FilteredStream::evaluateBoolean

The 'select_node' pointer was not initialized on all code paths, leading to the risk of reading it before assigning a value. This could cause undefined behavior.

Added initialization of the pointer to 'nullptr' when it is declared to fix the bug.

laBobberto committed:
fix Initialize variables in convertGregorianDateToWeekDate

The variables 'yearNumber' and 'weekNumber' were not initialized when declared. Due to complex conditional logic, the compiler issued a warning about the possible use of an uninitialized variable 'weekNumber'.

Added zero initialization to eliminate the warning.

laBobberto committed:
fix: Initialize variables in BTR_reserve_slot

The 'len' and 'space' variables were not initialized when declared, causing a compiler warning.

Added zero initialization to eliminate the warning and improve code reliability.

laBobberto committed:
fix: Initialize multiKeyLength in compress function

The 'multiKeyLength' variable was only initialized when handling international string types, but was read for all types. This led to reading uninitialized memory and undefined behavior.

Added initialization 'multiKeyLength = 0' on declaration to ensure correct logic for all index types.

laBobberto committed:
fix: Fix use of uninitialized ptr and length in compress

The 'ptr' pointer and the 'length' variable were not initialized on all code paths inside the do-while loop, leading to reading from uninitialized memory and undefined behavior on the second and subsequent iterations.

Added initialization ('ptr = nullptr' and 'length = 0') to fix these bugs and eliminate compiler warnings.

XaBbl4 commented on src/jrd/filters.cpp:
```diff
@@ -1044,7 +1044,7 @@ ISC_STATUS filter_transliterate_text(USHORT action, BlobControl* control)
 	// Do we already have enough bytes in temp buffer to fill output buffer?
 
-	bool can_use_more;
+	bool can_use_more = true;
```
Просто инициализировать переменную без переделки остального кода - это совершенно не оптимально.

Тут можно было проанализировать код дальше и понять, что данный report ошибочный, переменная `can_use_more` всегда инициализируется внутри условия `if (length)`, а в случаях когда `length == 0`, тогда и условие `if (!length || can_use_more)` всегда будет проходить по первому его предикату, следовательно проверка никогда не дойдет до неинициализированной переменной `can_use_more`.

Тогда этот кусок кода можно было переделать, например, следующим образом:
```c++
bool can_use_more = true;
USHORT length = aux->ctlaux_buffer1_unused;
// Always keep a minimal count of bytes in the input buffer,
// to prevent the case of truncated characters.
if (length >= 4 && control->ctl_buffer_length < (length * aux->ctlaux_expansion_factor / EXP_SCALE))
{
	// No need to fetch more bytes, we have enough pending
	can_use_more = false;
}
```
и из условия `if (!length || can_use_more)` убрать первый предикат, оставив только `can_use_more`, т.к. в случае если выполнение не попадет в переделанное условие, то переменная `can_use_more` будет иметь истинное значение и выполнение все равно попадет внутрь условия, ну и можно убрать закомментированное перед этим старое условие, которые было добавлено 17 лет назад, как явно уже ненужный участок кода.

laBobberto replied:
Переделал код примерно, как вы и сказали

XaBbl4 replied:
Вижу, но у меня был сохранен комментарий по поводу минимального количества байт, вы точно считаете, что он зря там оставлен и теперь больше не нужен, но при этом старое закомментированное условие все еще актуально?

laBobberto replied:
Не подумал, комментарий добавил, вот коммит: 8f23b0a

XaBbl4 commented on src/gpre/pat.cpp:
```diff
@@ -179,9 +179,9 @@ void PATTERN_expand( USHORT column, const TEXT* pattern, PAT* args)
 	p += align(p, column);
 
-	SSHORT value;				// value needs to be signed since some of the
+	SSHORT value = 0;				// value needs to be signed since some of the
```
Тут при добавлении инициализации уезжает комментарий, вернуть бы его на место.

Мельком глянул код этой процедуры и что-то он совсем древний и требует рефакторинга, например, `sw_ident` всегда выставляется вместе с `long_flag`, следовательно, можно сократить код, и возможно эти переменные вообще ликвидируются, но это вряд ли задача этого PR

laBobberto replied:
Комментарий поправил

XaBbl4 replied:
Хорошо

XaBbl4 commented on src/dsql/StmtNodes.cpp:
```diff
@@ -521,7 +521,7 @@ BlockNode* BlockNode::pass2(thread_db* tdbb, CompilerScratch* csb)
 {
 	jrd_tra* transaction = request->req_transaction;
-	SavNumber savNumber;
+	SavNumber savNumber{};
```
Почему здесь и в других местах проинициализировано пустыми скобками?
А в некоторых местах явными значениями `0`, `0.0` и т.д.

laBobberto replied:
Численные переменные я целенаправленно объявляю константами, а структуры с помощью {}. Насколько мне известно {} объявляет поля структуры аналогичными константами. Если надо могу переделать все под единый формат.

XaBbl4 replied:
Просто `SavNumber` - это не структура, а `typedef SINT64` (что в свою очередь `typedef long long int`), объявленные в `fb_types.h`, просто стало интересно почему так, по сути - это равноценная инициализация и в переменной будет 0, но странно :-)

laBobberto replied:
Тоже вот не сразу посмотрел, но поправил. Просто инициализировать примитивные типы константами понятнее, нежели использовать {}, хотя он тоже это делает. Вот коммит: c6185ab

XaBbl4 commented on src/dsql/StmtNodes.cpp:
```diff
@@ -5067,7 +5067,7 @@ void ExecBlockNode::genBlr(DsqlCompilerScratch* dsqlScratch)
 	const bool subRoutine = dsqlScratch->flags & DsqlCompilerScratch::FLAG_SUB_ROUTINE;
 
-	unsigned returnsPos;
+	unsigned returnsPos = 0;
```
Тут тоже можно было проанализировать детальнее и понять, что `returnsPos` нужна только в случае когда `subRoutine == true`, следовательно инициализацию можно перенести ниже по коду и убрать присвоение в теле условия `if (!subRoutine)`

laBobberto replied:
Не совсем вас как переписать функцию. returnsPos используется в блоке if (!subRoutine) для определения позиций выходных данных (насколько я понял). Если бы ее объявили только в if (subRoutine), то код в if (!subRoutine) сломался бы, потому что returnsPos была бы вне зоны видимости.

XaBbl4 replied:
> returnsPos используется в блоке if (!subRoutine)

Оно там не используется, а ему присваивается значение `dsqlScratch->variables.getCount();`, но по факту - это не нужно, т.к. при условии `if (!subRoutine)` оно никогда не будет использовано далее в коде, т.к. переменная используется только в цикле `for (unsigned i = 0; i < returnsPos; ++i)` внутри условия `if (subRoutine)`, также в аналогичном условии чуть выше ему присваивается значение `dsqlScratch->variables.getCount() - dsqlScratch->outputVariables.getCount()`, которое как раз и будет использовано в условии цикла.

Следовательно, можно убрать присвоение внутри `if (!subRoutine)`, и перенести инициализацию ниже, либо перед первым условием `if (subRoutine)`, либо вообще перед циклом внутри условия, но это нужно чуть внимательнее посмотреть: "могут ли измениться переменные `dsqlScratch->variables` и `dsqlScratch->outputVariables` между теми двумя условиями, где присваивается значение переменной и где оно используется..."

laBobberto replied:
Поправил тут: 1519ed2

XaBbl4 commented on src/dsql/StmtNodes.cpp:
```diff
@@ -12195,7 +12195,7 @@ static void validateExpressions(thread_db* tdbb, const Array<ValidateInfo>& vali
 		{
 			// Validation error -- report result
-			const char* value;
+			const char* value = nullptr;
```
Ок, инициализация этой переменной поможет закрыть этот report, но появится другой, в котором будет неопределенное поведение, т.к. анализатор увидит `nullptr[lenght] = 0`

Это, на самом деле, тоже ошибочный report, т.к. все случае покрываются предыдущим кодом и не может возникнуть ситуации когда в это место придет неинициализированная переменная, думаю, добавление ассерта перед преобразованием должно исправить проблему.

laBobberto replied:
Пофиксил в 3cc0fb2

XaBbl4 replied:
Хорошо

XaBbl4 commented on src/common/unicode_util.cpp:
```diff
@@ -1853,7 +1853,7 @@ USHORT UnicodeUtil::Utf16Collation::stringToKey(USHORT srcLen, const USHORT* src
 				{
 					UCHAR lastCharKey[BUFFER_TINY];	// sort key for a single character
-					ULONG prefixLen, lastCharKeyLen;
+					ULONG prefixLen = 0, lastCharKeyLen = 0;
```
При добавлении инициализации переменной prefixLen можно тогда сразу убрать блок:
```c++
else
	prefixLen = 0;
```

laBobberto replied:
Поправил в e7fc91b

XaBbl4 replied:
Хорошо

XaBbl4 commented on src/jrd/sort.cpp:
```diff
@@ -417,7 +417,7 @@ void Sort::sort(thread_db* tdbb)
 	run_control* run;
 	merge_control* merge;
-	merge_control* merge_pool;
+	merge_control* merge_pool = nullptr;
```
Тут явно древний c-style код, и простая инициализация nullptr тут выглядит глупо, т.к. это тупо закрывает отчет анализатора, и совсем никак не улучшает сам код.

`merge_pool` инициализируется в одном месте при условии `count > 1` и используется тоже в одном месте в цикле с условием `count > 1`, следовательно можно переписать код таким образом, чтобы объявление и инициализация была внутри этого условия `count > 1`

laBobberto replied:
Поправил в d0d88ab

XaBbl4 replied:
Хорошо

XaBbl4 commented on src/utilities/guard/guard.cpp:
```diff
@@ -183,7 +183,7 @@ int CLIB_ROUTINE main( int argc, char **argv)
 
 	do {
-		int ret_code;
+		int ret_code = 0;
```
Во-первых, объявление `ret_code` можно переместить ниже по коду, во-вторых, присвоить ему константу `NORMAL_EXIT`, раз уж она используется далее по коду.

laBobberto replied:
Поправил в 0284e5a

XaBbl4 replied:
Когда я говорил ниже по коду, то имел ввиду оставить внутри цикла, но поместить перед объявлением `shutdown_child`, а не вынести из цикла. Вне цикла эта переменная не нужна, т.к. нигде дальше не используется.

laBobberto replied:
Исправил, согласно вашим замечаниям. Коммит: 108d0fa

XaBbl4 replied:
Отступ сделан пробелами вместо табуляции.

laBobberto replied:
Тут исправил 9464971

XaBbl4 commented on src/jrd/recsrc/FilteredStream.cpp:
```diff
@@ -187,7 +187,7 @@ bool FilteredStream::evaluateBoolean(thread_db* tdbb) const
 
 	// ANY/ALL select node pointer
-	const BoolExprNode* select_node;
+	const BoolExprNode* select_node = nullptr;
```
Также при этом можно убрать блок:
```c++
else
	select_node = NULL;
```

laBobberto replied:
Поправил в 7abf372

XaBbl4 replied:
Ок

XaBbl4 commented on src/common/classes/NoThrowTimeStamp.cpp:
```diff
@@ -356,7 +356,7 @@ int NoThrowTimeStamp::convertGregorianDateToWeekDate(const struct tm& times) noe
 
 	// Find if y m d falls in yearNumber y-1, weekNumber 52 or 53
-	int yearNumber, weekNumber;
+	int yearNumber = 0, weekNumber = 0;
```
Здесь report был только на `weekNumber`, т.к. `yearNumber` в любом случае инициализируется в теле следующего условия.
Ну и судя по функции, она не предполагает, что `weekNumber` может быть равен 0, вернуться должно значение от 1 до 53, поэтому данная инициализация бестолковая.

`weekNumber` кстати, в текущем коде, также не остался бы без инициализации, но анализатор не справился и выдал очередной ложный отчет, тут можно переместить последнее условие внутрь блока else, т.к. только в одном случае может быть ситуация, когда `yearNumber == y` и тогда, возможно, анализатор сможет справиться и не нужно будет добавлять инициализацию при объявлении.

laBobberto replied:
Поправил в 8f70f84

XaBbl4 replied:
Хорошо, при этом исправился отчет в статическом анализаторе?

laBobberto replied:
Не, но вот при таком изменение 2cc6b6f все ушло. Изменил логику присваивания.

XaBbl4 replied:
Сходимость результатов до и после не сравнивал, но что сразу бросилось в глаза, так это зачем в такой реализации нужна переменная `yearNumber`, мы ее не возвращаем и используем только в одном месте (где просто можно заменить на `y-1`), ранее она использовалась в условии "не изменился ли `yearNumber` при пограничном случае, когда год увеличивается на единицу, а номер недели сбрасывается в единицу", что тоже можно было изменить, добавив блок else и убрав эту переменную вообще.

Ранее для пограничных случаев (когда `weekNumber` возвращался либо 1 либо 52/53) задействовалось меньше операций, т.к. сначала проверялись они, и только в ином случае вычислялся возвращаемый номер недели, а сейчас этот кусок вынесен за пределы проверок и всегда вычисляется, а потом только проверяется "а может это пограничный случай?" и в итоге затираются предыдущие вычисления.

Просто может быть такая ситуация, что в ходе избавления от безобидного ложного предупреждения, мы снизим производительность, поэтому хорошо бы замерить ее до изменений, и после, на разных тестовых наборах с 1 млн/млрд вызовов.
Т.е. нужно рефакторить код таким образом, чтобы результат функции на всем множестве входных данных остался таким же, но при этом и производительность не пострадала, а может даже и улучшилась.

laBobberto replied:
Насчёт переменной - ее убрать не трудно (насколько я понимаю). Я могу сделать тесты и собрать данные с помощью какого-нибудь perf stat, после построить графики. Но это будет в в формате вызова немного переделанной функции и проверки ее данных из какого-нибудь питона - вас устроит такой вариант?

XaBbl4 replied:
Тут главное результат, чтобы убедиться, что функция не стала работать медленнее, чем работала до такого рефакторинга. Если графики это смогут показать - хорошо.
Но я бы перед этим подумал как можно еще лучше переделать функцию, чтобы и статический анализатор не ругался и результат остался таким же и замерять производительность, возможно, не потребовалось бы, т.к. изначальная логика сохранилась бы неизменной. Могу дать подсказку, если нет желания самому дойти до этого?

laBobberto replied:
Давайте я пока сам попытаюсь, хотелось бы дойти самому

laBobberto replied:
Решил использовать ранний выход и битовые операции. Вот сама функция: 5a10754. На кодчекере не тестил, но он же вряд-ли будет ругаться на ранние выходы. Вот здесь мой небольшой ресерч(в том же репозитории и даные и тестер и функции на тест): [https://github.com/laBobberto/ConvertData/blob/master/benchmark_summary_charts.png](url)
Я тестил на 4 компиляторах, хотя можно было на двух. Все билдилось на стандартной оптимизации. Из-за приколов таймера виндовс (с большим размером тестовых данных в таблицы не записывалось медианное, минимальное время) были сделаны тесты с батчингом и уменьшенным наборам данных. Если необходимо, то могу изменить методологию тестирования и расширить промежуток годов. Тесты в нынешний момент проводились с такими данными: кол-во итераций - 1000000000, размер тестовых данных - 50000000, года - от 1800 до 3000

XaBbl4 replied:
Хорошо, я собственно только про ранний выход и имел ввиду в предыдущем своем сообщении, расставив их, предполагаю можно было избавиться от ошибок анализатора и сохранить весь код без изменений, из-за чего не требовалось бы его тестировать, т.к. гарантировано поведение бы не изменилось. Но сейчас гораздо больше оптимизаций сделано, которые еще и доказаны тестированием. Отлично!

XaBbl4 commented on src/jrd/btr.cpp:
```diff
@@ -2315,7 +2315,7 @@ void BTR_reserve_slot(thread_db* tdbb, IndexCreation& creation)
 
 	UCHAR* desc = 0;
-	USHORT len, space;
+	USHORT len = 0, space = 0;
```
И тут анализатор какой-то глупый, мы в любом случае хотя бы раз, но попадем в цикл `for (int retry = 0; retry < 2; ++retry)` и тут же проинициализируем обе эти переменные, следовательно не сможем перепрыгнуть цикл и использовать неинициализированные переменные.
Не вижу смысла их сначала инициализировать 0, а потом в цикле нужным значением, что нам мешает сразу проинициализировать правильно и переделать их присвоение в цикле?

laBobberto replied:
Поправил в 687c238. Занес объявление переменных в for. По идее каждую итерацию в цикле им задается новое значение вне зависимости от предыдущих.

XaBbl4 replied:
Ох-ох-ох... И что, не возникло ошибки компиляции? ;-)

laBobberto replied:
Была... Поэтому присваивание было перенесено в конец цикла, а их инициализация вне цикла(4a7916a), т.к. они использовались ниже.

XaBbl4 replied:
А тут `len` разве может измениться внутри цикла и его точно нужно сбрасывать в конце?

laBobberto replied:
Честно, не знаю. Просто изначально в цикле было постоянное обновление len, и я решил не отходить от традиций. Ну я поправлю это

XaBbl4 replied:
Тут нет никаких традиций, код был написан десятки лет назад, и несколько раз переписывался, возможно ранее она действительно требовала обновления внутри цикла, или может так было удобно программисту, который его изначально написал.
Сейчас читая код мы видим, что она не меняется, возможно, это сможет исправить сам оптимизатор компилятора, но может и нет, следовательно зачем тратить на это процессорное время, также для лучшей читаемости кода ее можно объявить как константу и тогда у следующего человека, который будет это читать будет возникать меньше вопросов.

laBobberto replied:
Поправил, сделал переменную константой, убрал постоянное присваивание в цикле. Все тут: b122e4d

laBobberto commented:
Thanks for the detailed review! I realized that simply adding initialization isn't the best approach. I'll analyze each case and revise the PR based on your comments.

XaBbl4 commented:
You can make corrections in the current PR, then the comment history will be saved and it will be easier to review in the future.

laBobberto committed:
fix: Refactor filters.cpp logic for 'can_use_more' and remove redundant checks

laBobberto committed:
fix: Initialize value in PATTERN_expand and align comment (gpre/pat.cpp)

laBobberto committed:
fix: add assert 'value'

laBobberto committed:
fix: Narrow scope of merge_pool pointer in Sort::sort method

laBobberto committed:
fix: Scope ret_code outside of loop and initialize with NORMAL_EXIT in main

laBobberto committed:
fix: Initialize select_node in FilteredStream::evaluateBoolean and remove redundant else block

laBobberto committed:
fix: Refactor weekDate logic to remove unnecessary initialization and false positive on weekNumber

laBobberto committed:
refactor: Narrow scope of 'len' and 'space' in BTR_reserve_slot to eliminate redundant initialization

laBobberto committed:
refactoring: remove unnecessary else

laBobberto committed:
refactoring: while and if blocks using merge_pool

XaBbl4 commented:
А вообще после всех сделанных исправлений, повторно запускался статический анализ кода? Эти отчеты в нем исправились и при этом не появилось новых?

Просто судя по переносу переменной `space` внутрь цикла, то думаю, что сборку даже не запускали, т.к. эта переменная, насколько вижу по коду, используется вне цикла, а ее объявление теперь внутри цикла.

laBobberto commented:
Вы правы, похоже я сравнил не те отчёты в кодчекере

laBobberto committed:
fix: initialization custom types in 'StmtNodes.cpp'.

laBobberto committed:
fix: moving the preassignment of 'len', 'space' to the end of the for iterations and initializing them from the beginning with the correct values

laBobberto committed:
fix: Initializing yearNumber with y and weekNumber with computed value instead of declaring them uninitialized. Move the condition (yearNumber == y) inside the else block since it can only be true in that case. Remove intermediate variable j.

laBobberto committed:
refactor: move ret_code declaration into loop scope

laBobberto committed:
refactor: remove old comment and add new

laBobberto committed:
refactor: move returnsPos declaration to its actual usage point inside the subRoutine validation block.

laBobberto commented:
Проект собирал, вроде все собралось. Ошибок в кодчекере не добавилось

XaBbl4 commented:
Посмотрел последние изменения, все ок, только в двух местах остались замечания, которые стоит проверить.

laBobberto committed:
refactor: indentation fixed

laBobberto committed:
refactor: optimize week date conversion with bitwise operations
1) Use bit shifts instead of division by 4
2) Replace % 4 with bitwise AND
3) Use bitwise &/| instead of logical &&/|| for branchless code

laBobberto committed:
refactor: the type of the variable `len` has been changed to const USHORT, and the constant initialization with the same value has been removed.

laBobberto commented:
Получается, что я все исправил. Моя работа в этом пулреквесте окончена?

XaBbl4 commented:
Да, вроде все замечания поправлены, хорошо справился!

laBobberto commented:
> Да, вроде все замечания поправлены, хорошо справился!

Тогда спасибо вам за детальные кодревью
